sdk/core/azure-core/azure/core/tracing/context.py

```diff
@@ -0,0 +1,161 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import threading
+
+try:
+    from typing import TYPE_CHECKING
+except ImportError:
+    TYPE_CHECKING = False
+
+if TYPE_CHECKING:
+    from typing import Any, Callable
+    from typing_extensions import Protocol
+else:
+    Protocol = object
+
+try:
+    import contextvars
+except ImportError:
+    contextvars = None
+
+
+class ContextProtocol(Protocol):
+    """
+     Implements set and get variables in a thread safe way.
+     """
+
+    def __init__(self, name, default, lock):
+        # type: (string, Any, threading.Lock) -> None
+        pass
+
+    def clear(self):
+        # type: () -> None
+        """Reset the value to the default value"""
+        pass
+
+    def get(self):
+        # type: () -> Any
+        """Get the stored value."""
+        pass
+
+    def set(self, value):
+        # type: (Any) -> None
+        """Set the value in the context."""
+        pass
+
+
+class _AsyncContext:
+    """
+    Uses contextvars to set and get variables globally in a thread safe way.
+    """
+
+    def __init__(self, name, default, lock):
+        self.name = name
+        self.contextvar = contextvars.ContextVar(name)
+        self.default = default if callable(default) else (lambda: default)
+        self.lock = lock
+
+    def clear(self):
+        # type: () -> None
+        """Reset the value to the default value"""
+        self.contextvar.set(self.default())
+
+    def get(self):
+        # type: () -> Any
+        """Get the stored value."""
+        try:
+            return self.contextvar.get()
+        except LookupError:
+            value = self.default()
+            self.set(value)
+            return value
+
+    def set(self, value):
+        # type: (Any) -> None
+        """Set the value in the context."""
+        with self.lock:
+            self.contextvar.set(value)
+
+
+class _ThreadLocalContext:
```

Comment:
Is omitting `object` base class here and below intentional?

Reply:
Was not intentional. Will add object as a base class.